timemachine/fe/free_energy.py

```diff
@@ -266,6 +264,45 @@ def boxes(self) -> List[NDArray]:
         return [np.array(traj.boxes) for traj in self.trajectories]
 
 
+@dataclass
+class HREXSimulationResult(SimulationResult):
+    hrex_diagnostics: HREXDiagnostics
+    hrex_plots: HREXPlots
+
+    def extract_trajectories_by_replica(self, atom_idxs: NDArray) -> NDArray:
+        """Returns an array of shape (n_replicas, n_frames, len(atom_idxs), 3) of trajectories for each replica
+
+        Note: This consumes O(n_frames * len(atom_idxs)) memory, and thus may OOM for large systems if len(atom_idxs) is
+        a significant fraction of the total number of atoms.
+
+        Parameters
+        ----------
+        atom_idxs: NDArray
+            Indices of atoms to extract
+        """
+
+        # (states, frames, atoms, 3)
+        trajs_by_state = np.array(
+            [
+                np.concatenate([np.array(chunk)[:, atom_idxs] for chunk in state_traj.frames._chunks()], axis=0)
+                for state_traj in self.trajectories
+            ]
+        )
+
+        replica_idx_by_iter_by_state = np.asarray(self.hrex_diagnostics.replica_idx_by_state_by_iter).T
+        state_idx_by_iter_by_replica = np.argsort(replica_idx_by_iter_by_state, axis=0)
+
+        # (replicas, frames, atoms, 3)
+        trajs_by_replica = np.take_along_axis(trajs_by_state, state_idx_by_iter_by_replica[:, :, None, None], axis=0)
+
+        return trajs_by_replica
+
+    def extract_ligand_trajectories_by_replica(self):
+        """Returns an array of shape (n_replicas, n_frames, n_ligand_atoms, 3) of ligand trajectories for each replica"""
+        ligand_idxs = self.final_result.initial_states[0].ligand_idxs
```

Comment:
[paranoia] currently (and for the foreseeable future), `s.ligand_idxs` should be identical `for all s in initial_states` -- but should we assert this just in case?

Reply:
Good call. Added assertion in https://github.com/proteneer/timemachine/pull/1320/commits/892cca96bcc41fbccdeb20ab7851e86f18ae8bff